node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1947 @@
+			} else {
+				modify_reputation(ctx, remote, COST_MESSAGE_NOT_DECODABLE).await?;
+			}
+		}
+	}
+	Ok(())
+}
+
+fn derive_erasure_chunks_with_proofs(
+	n_validators: usize,
+	validator_index: ValidatorIndex,
+	available_data: &AvailableData,
+) -> Result<Vec<ErasureChunk>> {
+	if n_validators <= validator_index as usize {
+		warn!(target: TARGET, "Validator index is outside the validator set range");
+		return Err(Error::Logic);
+	}
+	let chunks: Vec<Vec<u8>> = if let Ok(chunks) = obtain_chunks(n_validators, available_data) {
+		chunks
+	} else {
+		warn!(target: TARGET, "Failed to create erasure chunks");
+		return Err(Error::Logic);
+	};
+
+	// create proofs for each erasure chunk
+	let branches = branches(chunks.as_ref());
+
+	let erasure_chunks = branches
+		.enumerate()
+		.map(|(index, (proof, chunk))| ErasureChunk {
+			chunk: chunk.to_vec(),
+			index: index as _,
+			proof,
+		})
+		.collect::<Vec<ErasureChunk>>();
+
+	Ok(erasure_chunks)
+}
+
+/// Handle the changes necessary when our view changes.
+async fn handle_our_view_change<Context>(
+	ctx: &mut Context,
+	keystore: KeyStorePtr,
+	state: &mut ProtocolState,
+	view: View,
+) -> Result<()>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let old_view = std::mem::replace(&mut (state.view), view);
+
+	// needed due to borrow rules
+	let view = state.view.clone();
+	let added = view.difference(&old_view).collect::<Vec<&'_ Hash>>();
+
+	// add all the relay parents and fill the cache
+	for added in added.iter() {
+		let added = **added;
+		let validators = query_validators(ctx, added).await?;
+		let validator_index = obtain_our_validator_index(
+			&validators,
+			keystore.clone(),
+		);
+		state.add_relay_parent(ctx, added, validators, validator_index).await?;
+	}
+
+	// handle all candidates
+	for (candidate_hash, _receipt) in state.cached_live_candidates_unioned(added) {
+		let per_candidate = state
+			.per_candidate
+			.entry(candidate_hash)
+			.or_default();
+
+		// assure the node has the validator role
+		let validator_index = if let Some(validator_index) = per_candidate.validator_index {
+			validator_index
+		} else {
+			continue;
+		};
+
+		// pull the proof of validity
+		let available_data = if let Some(available_data) =
+			query_available_data(ctx, candidate_hash.clone()).await?
+		{
+			available_data
+		} else {
+			continue;
+		};
+
+
+		let erasure_chunks = if let Ok(erasure_chunks) = derive_erasure_chunks_with_proofs(
+			per_candidate.validators.len(),
+			validator_index,
+			&available_data,
+		) {
+			erasure_chunks
+		} else {
+			return Ok(());
+		};
+
+		if let Some(erasure_chunk) = erasure_chunks.get(validator_index as usize) {
+			match store_chunk(ctx, candidate_hash, validator_index, erasure_chunk.clone()).await {
+				Err(e) => warn!(target: TARGET, "Failed to send store message to overseer: {:?}", e),
+				Ok(Err(())) => warn!(target: TARGET, "Failed to store our own erasure chunk"),
+				Ok(Ok(())) => {}
```

Comment:
Again, if you do a `QueryChunk` it will make it available. I think that in the end you can completely remove the logic with `obtain_chunks` and the dependency on `polkadot-erasure-encoding`.